Assets/Scripts/Switch.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class Switch : MonoBehaviour
+{
+    [SerializeField]
+    [Tooltip("The color of the switch. 0 if yellow, 1 if blue, 2 if red, 3 if green.")]
+    [Range(0, 3)]
+    private int switchColor;
+
+    [SerializeField]
+    [Tooltip("The puzzle this switch is connected to.\nMust have a script which implements the IPuzzle interface.")]
+    private GameObject connectedPuzzle;
+
+    [SerializeField]
+    [Tooltip("The character this switch adds to the solution.")]
+    private char solutionEntry = '1';
+
+    /// <summary>
+    /// The animator component attached to the switch
+    /// </summary>
+    private Animator animator;
+
+    /// <summary>
+    /// Is the switch active or not?
+    /// </summary>
+    private bool isSwitchOn;
+
+    /// <summary>
+    /// The IPuzzle component of the connected puzzle
+    /// </summary>
+    private IPuzzle puzzle;
+
+    // Start is called before the first frame update
+    void Start()
+    {
+        animator = GetComponent<Animator>();
+        animator.SetInteger("Color", switchColor);
+        puzzle = connectedPuzzle.GetComponent<IPuzzle>();
+        puzzle.AppendSwitchList(this);
+    }
+
+    /// <summary>
+    /// Sends the switch's solution to the puzzle and turns it on
+    /// Called by OnUse from the Dialogue System Usable component
+    /// </summary>
+    public void SetSwitch()
+    {
+        if (!isSwitchOn)
+        {
+            AnimateSwitch();
+            puzzle.AppendSolution(solutionEntry);
+        }
+    }
+
+    /// <summary>
+    /// Resets the switch to off
+    /// Called by Reset from the connected puzzle
+    /// </summary>
+    public void ResetSwitch()
+    {
+        if (isSwitchOn)
+        {
+            AnimateSwitch();
+        }
+    }
+
+    /// <summary>
+    /// Store the switch state and trigger animations
+    /// </summary>
+    private void AnimateSwitch()
```

Comment:
SetSwitch, AnimateSwitch, and ResetSwitch are confusing. Figure out a way to clarify this.

Basically, any time the switch state changes (on or off) you want to animate the switch, right? So maybe make that a private property bool IsSwitchOn and inside it's setter you set the bool anim param to match the new value of isSwitchOn. That gets rid of the confusing AnimateSwitch function.

SetSwitch is confusing because it implies the user (programmer calling method) can pass in the new state (set it). But you can't it doesn't seem. I really don't follow what it means to "set" a switch in this context.

ResetSwitch only makes sense to me in the context of, the switch could start on or off, based on how the designer sets it up originally, and maybe we need to remember that in case we need to reset the switch back to it's original status. But that's not what it looks like the method actually does at all.